### PR TITLE
ENHANCE: Task state change modals

### PR DESCRIPTION
**Pull request by TejaswiniMannuru24**
@jakerenzella When a student wants to move to previous states (not started, working on it, need help) from ready for feedback before the task's deadline, the below confirmation box will be displayed
before deadline

But if the same action is performed after deadline, the following warning box would be displayed
after deadline

### CODE REVIEW

File: src/app/projects/states/dashboard/directives/task-dashboard/directives/task-status-card/task-status-card.coffee [hunk 1]
```diff
@@ -43,10 +43,24 @@ angular.module('doubtfire.projects.states.dashboard.directives.task-dashboard.di
...
       else
-        $scope.task.triggerTransition(trigger, $scope.unitRole)
+        $scope.triggerStatusChange(trigger)
 
     # Allow upload of new files for an updated submission
     $scope.updateFilesInSubmission = ->
       taskService.presentTaskSubmissionModal($scope.task, $scope.task.status, true)
 
+    $scope.triggerStatusChange = (trigger) ->
+      if (trigger == 'not_started' || trigger == 'working_on_it' || trigger == 'need_help') && $scope.task.status == 'ready_to_mark' && !$scope.task.isOverdue()
```
Comment (jakerenzella) [inline src/app/projects/states/dashboard/directives/task-dashboard/directives/task-status-card/task-status-card.coffee]: We should have helper methods to determine this case, we don't want to have the state checks in the component like this, check the taskService and the `isInxxxState` methods.

File: src/app/projects/states/dashboard/directives/task-dashboard/directives/task-status-card/task-status-card.coffee [hunk 2]
```diff
@@ -43,10 +43,24 @@ angular.module('doubtfire.projects.states.dashboard.directives.task-dashboard.di
...
+        ConfirmationModal.show(
+          "Do you want to change the status?",
+          "Changing the status will cancel all previous submissions.
+          Use confirm to change the status.",
+          () -> $scope.task.triggerTransition(trigger, $scope.unitRole))
+      else if (trigger == 'not_started' || trigger == 'working_on_it' || trigger == 'need_help') && $scope.task.status == 'ready_to_mark' && $scope.task.isOverdue()
```
Comment (jakerenzella) [inline src/app/projects/states/dashboard/directives/task-dashboard/directives/task-status-card/task-status-card.coffee]: Likewise here

File: src/app/projects/states/dashboard/directives/task-dashboard/directives/task-status-card/task-status-card.tpl.html [hunk 3]
```diff
@@ -1,5 +1,6 @@
 <div class="card card-default card-sm card-status">
   <div class="card-heading task-trigger-dropdown" ng-show="triggers.length > 0">
+    <!-- {{task.isPastTargetDate()}} -->
```
Comment (jakerenzella) [inline src/app/projects/states/dashboard/directives/task-dashboard/directives/task-status-card/task-status-card.tpl.html]: Should this be removed?

File: src/app/projects/states/dashboard/directives/task-dashboard/directives/task-status-card/task-status-card.coffee [hunk 4]
```diff
@@ -43,10 +43,24 @@ angular.module('doubtfire.projects.states.dashboard.directives.task-dashboard.di
...
+        WarningModal.show(
+          "Not allowed to perform this action",
+          "The deadline for this task has passed and you are not allowed to change status.
```
Comment (jakerenzella) [inline src/app/projects/states/dashboard/directives/task-dashboard/directives/task-status-card/task-status-card.coffee]: "The deadline for this task has passed and you are not allowed to change *the* status."

**Comment by macite** [comment]
Should we not allow them to cancel and re-upload a submission? In general we have always permitted actions, rather than blocking them.

File: src/app/projects/states/dashboard/directives/task-dashboard/directives/task-status-card/task-status-card.coffee [hunk 5]
```diff
@@ -43,10 +43,24 @@ angular.module('doubtfire.projects.states.dashboard.directives.task-dashboard.di
...
 
     # Allow upload of new files for an updated submission
     $scope.updateFilesInSubmission = ->
       taskService.presentTaskSubmissionModal($scope.task, $scope.task.status, true)
 
+    $scope.triggerStatusChange = (trigger) ->
+      if (trigger == 'not_started' || trigger == 'working_on_it' || trigger == 'need_help') && $scope.task.status == 'ready_to_mark' && !$scope.task.isOverdue()
+        ConfirmationModal.show(
+          "Do you want to change the status?",
+          "Changing the status will cancel all previous submissions.
```
Comment (macite) [inline src/app/projects/states/dashboard/directives/task-dashboard/directives/task-status-card/task-status-card.coffee]: "Changing the status will cancel the previous submission"

File: src/app/projects/states/dashboard/directives/task-dashboard/directives/task-status-card/task-status-card.tpl.html (hunk 3 above)
Comment (TejaswiniMannuru24) [reply src/app/projects/states/dashboard/directives/task-dashboard/directives/task-status-card/task-status-card.tpl.html]: Yes Jake..this has to be removed

**Comment by TejaswiniMannuru24** [comment]
> Should we not allow them to cancel and re-upload a submission? In general we have always permitted actions, rather than blocking them.

We are letting the student know that switching to previous states will remove existing submissions. So basically, if they move back to **"not started"** from **"Ready for feedback"**, the confirmation action should change the status to **"not started"** and not directly take them to re-upload submission. They can do this by clicking on Ready-for-feedback again.
May be the message **"Changing the status will cancel the previous submissions and you will have to re-submit by clicking on Ready-for-feedback again"** would resolve any confusion?

**Comment by macite** [comment]
I was thinking when it is overdue... doesn't this change block that possibility?

**Comment by TejaswiniMannuru24** [comment]
@macite to let the students re-submit past deadline, I have now removed the warning box. 
after deadline1

File: src/app/projects/states/dashboard/directives/task-dashboard/directives/task-status-card/task-status-card.coffee [hunk 6]
```diff
@@ -50,17 +50,19 @@ angular.module('doubtfire.projects.states.dashboard.directives.task-dashboard.di
       taskService.presentTaskSubmissionModal($scope.task, $scope.task.status, true)
 
     $scope.triggerStatusChange = (trigger) ->
-      if (trigger == 'not_started' || trigger == 'working_on_it' || trigger == 'need_help') && $scope.task.status == 'ready_to_mark' && !$scope.task.isOverdue()
+      if (trigger == 'not_started' || trigger == 'working_on_it' || trigger == 'need_help') && $scope.task.inAwaitingFeedbackState() && !$scope.task.isOverdue()
```
Comment (TejaswiniMannuru24) [inline src/app/projects/states/dashboard/directives/task-dashboard/directives/task-status-card/task-status-card.coffee]: @jakerenzella Have used helper methods for the task status check.

**Pull request closed without merging**